synapse/replication/tcp/streams/_base.py

```diff
@@ -153,61 +162,62 @@ def discard_updates_and_advance(self):
-                Resolves to a pair ``(updates, current_token)``, where ``updates`` is a
-                list of ``(token, row)`` entries. ``row`` will be json-serialised and
-                sent over the replication steam.
+            Resolves to a pair `(updates, new_last_token, limited)`, where
+            `updates` is a list of `(token, row)` entries, `new_last_token` is
+            the new position in stream, and `limited` is whether there are
+            more updates to fetch.
         """
-        updates, current_token = await self.get_updates_since(self.last_token)
+        current_token = self.current_token()
+        updates, current_token, limited = await self.get_updates_since(
+            self.last_token, current_token
+        )
         self.last_token = current_token
 
-        return updates, current_token
+        return updates, current_token, limited
 
     async def get_updates_since(
-        self, from_token: int
-    ) -> Tuple[List[Tuple[int, JsonDict]], int]:
+        self, from_token: Union[int, str], upto_token: int, limit: int = 100
+    ) -> Tuple[List[Tuple[int, JsonDict]], int, bool]:
         """Like get_updates except allows specifying from when we should
         stream updates
 
         Returns:
-            Resolves to a pair `(updates, new_last_token)`, where `updates` is
-            a list of `(token, row)` entries and `new_last_token` is the new
-            position in stream.
+            Resolves to a pair `(updates, new_last_token, limited)`, where
+            `updates` is a list of `(token, row)` entries, `new_last_token` is
+            the new position in stream, and `limited` is whether there are
+            more updates to fetch.
         """
 
         if from_token in ("NOW", "now"):
-            return [], self.current_token()
-
-        current_token = self.current_token()
+            return [], upto_token, False
 
         from_token = int(from_token)
 
-        if from_token == current_token:
-            return [], current_token
-
-        rows = await self.update_function(
-            from_token, current_token, limit=MAX_EVENTS_BEHIND + 1
-        )
-
-        # never turn more than MAX_EVENTS_BEHIND + 1 into updates.
-        rows = itertools.islice(rows, MAX_EVENTS_BEHIND + 1)
-
-        updates = [(row[0], row[1:]) for row in rows]
-
-        # check we didn't get more rows than the limit.
-        # doing it like this allows the update_function to be a generator.
-        if len(updates) >= MAX_EVENTS_BEHIND:
-            raise Exception("stream %s has fallen behind" % (self.NAME))
-
-        # The update function didn't hit the limit, so we must have got all
-        # the updates to `current_token`, and can return that as our new
-        # stream position.
-        return updates, current_token
+        if from_token == upto_token:
+            return [], upto_token, False
+
+        if self._is_worker and self._QUERY_MASTER:
```

Comment:
I can't help but feel that this would be much easier to follow if it were implemented in the individual Streams:

```
if hs.config.worker_app is None:
    self.update_function = handler.get_updates
else:
    self.update_function = <something to do with ReplicationGetStreamUpdates>
```

Reply:
This turns out to be surprisingly faffy, given that what the API call returns is not quite the same as what the `update_function` returns (as can be seen from the differences in the two branches below the commented line). I guess we could make them be identical (i.e. make the API hit only return the rows), but I'm not sure that really makes things clearer. Thoughts?

Reply:
Make the spec for `update_function` be that it return the same `(updates, new_last_token, limited)` tuple.

Have a helper function:
 
```python
def db_query_to_update_function(query_function):
    """Wraps a db query function which returns a list of rows to make it suitable for use as an `update_function` for the Stream class"""
    async def update_function(from_token, upto_token, limit):
            rows = await query_function(from_token, upto_token, limit=limit)
            updates = [(row[0], row[1:]) for row in rows]
            limited = False
            if len(updates) == limit:
                upto_token = rows[-1][0]
                limited = True

            return updates, upto_token, limited
    return update_function
```

Do something similar with the `ReplicationClient`.

Use that to wrap the db query function in the Stream class:

```python
        self.update_function = db_query_to_update_function(store.get_all_new_backfill_event_rows)
```

Suggest sorting out the type annotations to help with that.

I think it will be clearer than adding more and more fields to the `Stream` class to control how it is supposed to work.

Reply:
Ok, let me give that a go. (Annoyingly mypy refuses to accept setting functions like we do in stream, hence the `type: ignore` going on, which is apain)